### DISCUSSION TITLE
Use of QImg prop src="~assets" to access files in assets folder: Available elsewhere, e.g. in QBtn?

### DISCUSSION THREAD

**Discussion by mickey-bimlive** [quasarframework/quasar#12548] [General - Components / Directives / etc] [answered]
Recently I discovered the following powerful capability of QImg: In the QImg src prop, you can specifiy  a special "~" prefix, like in:
`<q-img src="~/assets/my-img.gif"/>` ("~assets/my-img.gif" works as well) to display images stored in the **assets folder**. My understanding: The assets folder is **protected,** while the statics folder is not. So if you happen to know or guess the name of a file in the statics folder, you can easily access the file in the browser through typing in its URL: `https://my.site.com/statics/my-img.gif`. This can be a security exposure. Access through URL will not work with e.g. `https://my.site.com/assets/my-img.gif`.` You will get an "Unauthorized" or "Cannot GET" response. So I consider the assets folder more secure. (Note: This is for an app that requires logon, only available to authorized users).

Now I'm in the process to move some of the static images of my site from the statics folder to the asssets folder, to get increased security. While doing this, I discovered that I also have some QBtns in my app, which use href props to display PDF docs (in a separate browser tab) located in a sub-folder /pdf under /statics:
`<q-btn type="a" target="_blank" href="https://my.site.com/pdfs/my-pdf.pdf"/>`

Now my question is whetherI could store those PDF docs under the assets folder rather than the statics folder to have them more secure, and still use QBtn to display theses PDF docs?

I didn't find any mentioning of the special "~" support in the QBtn props (this means href="~assets/..." will probably not work). So is this "~assets" to use the assets folder something special to QImg or is it a broader Vue or Quasar concept applicable also in other situations?

**Comment by MilosPaunovic** [comment]
Yes, you can place your PDFs into `assets` folder.
Give this a read, as it might be helpful: [https://quasar.dev/quasar-cli/handling-assets#introduction](https://quasar.dev/quasar-cli/handling-assets#introduction)

**Reply by mickey-bimlive** [reply]
@MilosPaunovic - the question is not only about PDFs, but whether QBtn with href to open an image or a PDF in a new browser tab is able to deal with the "tilde" ("~")to access the assets folder (rather than the statics folder), like QImg is. Here are examples (using an image rather than a PDF for convenience, but that shouldn't matter re. the "tilde"):  

Test 1: **QImg** with image at remote URL - **works**
`<q-img src="https://dummyimage.com/100.png"/> <!-- Works/displays remote image, inline -->`

Test 2: **QImg** with same image, now placed in src/assets folder - **works**
`<q-img src="statics/100.png"/> <!-- Works/displays same image now placed in statics, inline -->`

Test 3: **QImg** with same image, now placed in src/assets folder, using "~" syntax as described in Quasar docs - **works**
`<q-img src="~assets/100.png"/> <!-- Works/displays same image now placed in assets, inline  -->`

Test 4: Now using **QBtn** to display image at a remote URL in a separate browser tab - **works**
`<q-btn type="a" target="blank" href="https://dummyimage.com/100.png"/> <!-- Works/displays image in new browser tab -->`

Test 5: **QBtn** to display image placed in **src/statics** folder - this is what I also use for PDFs in statics! - **works**
`<q-btn type="a" target="blank" href="statics/100.png"/> <!-- Works/displays image placed in statics in new browser tab -->`

Test 6: **QBtn** to display image placed in **src/assets** folder, using "tilde" syntax per Quasar docs - **fails**
`<q-btn type="a" target="_blank" href="~assets/100.png"/> <!-- Syntax from q-img in test 3 - fails with 'Cannot GET' -->`
`<q-btn type="a" target="_blank" :href="'~assets/100.png'"/> <!-- Alt syntax from doc above - fails with 'Cannot GET' -->`
None of the last two "test 6" QBtn variants work - "Cannot GET" (= unauthorized) is shown in new browser tab both cases, and the browser address line shows the "~", as an indication that it is not being resolved correctly to point to the assets folder rather than statics folder, as it positively works with QImg src="...".

So, I'm wondering how to display an image (or alternatively a pdf) using q-btn in a separate browser tab, with an image (or pdf) in the assets folder (fails in test 6), not the statics folder (as it works in my app in general, and in test 5).

I understand that under the covers, q-btn href prop might be processed very differently from q-img src prop, but the Quasar doc advertises that the "~" as the only known means to access the assets folder should work broadly, and I'm not aware of a good other alternative to q-btn with href, to display something in a separate browser tab.

**Comment by mickey-bimlive** [comment, marked as answer]
After hours of experimenting, I found a solution addressing the earlier q-btn "fail" (test 6 above), enabling a q-btn with a href to a PDF (or other file) that resides in the protected assets folder, not accessible via regular URL in browser.

Let me share my solution for others. It comes with some complications and potential drawbacks....e.g. I believe the PDF is loaded into memory, not sure how that compares to regular hrefs:

1) Extend Webpack config in quasar.conf with support for PDF file loader:
```
extendWebpack(cfg, { isServer, isClient }) { 
cfg.module.rules.push(
      {
        test: /\.(png|svg|jpg|gif|pdf)$/,
        use: [
          {
            loader: 'file-loader',
            options: {
              name: '[name].[ext]'
            }
          }
        ]
      }
    )
...
}
```
2) In your app/page/component, include an import to the desired PDF file residing in or under the protected assets folder:
`import MyPDF from ("../assets/myPDF.pdf") `

3) In data(), add:
`myPDF: MyPDF,`

4) In computed, add a computed function that returns that PDF:
```
computeMyPDF() {
   return this.myPDF
 },
```

5) In the template section of your app/page/component, use that computed function in your  q-btn for the :href prop:
`<q-btn type="_a" :href="computeMyPDF"/>`

Clicking on that QBtn does the magic of opening the "protected" PDF in a separate browser tab! Without those steps, q-btn with a href pointing to a file in the assets folder fails with "Cannot GET..." (= test 6 above).

I would still be interested if there is an alternative solution, or even an option to use that similarly magic "~" support for files/images in the assets folder (available in q-img src prop) also with q-btn href prop :-)
